cumulusci/tasks/bulkdata/extract.py

```diff
@@ -205,27 +227,31 @@ def strip_name_field(record):
             )
         else:
             # If using the autogenerated id field, split out the returned records
-            # into two separate files and load into the main table and the sf_id_table
-
-            with tempfile.TemporaryFile("w+", newline="", encoding="utf-8") as f_values:
-                with tempfile.TemporaryFile(
-                    "w+", newline="", encoding="utf-8"
-                ) as f_ids:
-                    data_file_values, data_file_ids = self._split_batch_csv(
-                        record_iterator, f_values, f_ids
-                    )
-                    self._sql_bulk_insert_from_records(
-                        connection=conn,
-                        table=mapping.table,
-                        columns=columns[1:],  # Strip off the Id column
-                        record_iterable=csv.reader(data_file_values),
-                    )
-                    self._sql_bulk_insert_from_records(
-                        connection=conn,
-                        table=mapping.get_sf_id_table(),
-                        columns=["sf_id"],
-                        record_iterable=csv.reader(data_file_ids),
-                    )
+            # into two separate streams and load into the main table and the sf_id_table
+            values, ids = itertools.tee(record_iterator)
+            f_values = (row[1:] for row in values)
+            f_ids = (row[:1] for row in ids)
+
+            values_chunks = self._sql_bulk_insert_from_records_incremental(
+                connection=conn,
+                table=mapping.table,
+                columns=columns[1:],  # Strip off the Id column
+                record_iterable=f_values,
+            )
+            ids_chunks = self._sql_bulk_insert_from_records_incremental(
+                connection=conn,
+                table=mapping.get_sf_id_table(),
+                columns=["sf_id"],
+                record_iterable=f_ids,
+            )
+
+            def _consume(iterator):
+                # simplified from Python docs
+                collections.deque(iterator, maxlen=0)
+
+            # do the inserts one chunk at a time based on all of the
+            # generators nested previously.
+            _consume(zip(values_chunks, ids_chunks))
```

Comment:
I agree it's mostly a matter of taste. But please:
- Don't repeat the definition of _consume between here and utils.py
- Add a docstring that says what it does. The name "consume" is a clue but this linguistic weirdness was new and not very obvious to me, so I assume it would be to others too. Can you link to the spot in the Python docs that you simplified it from?

Reply:
68a5e62a7c4